kube/src/client/mod.rs

```diff
@@ -93,6 +114,34 @@ impl Client {
         Ok(res)
     }
 
+    /// Make WebSocket connection.
+    #[cfg(feature = "ws")]
+    pub async fn connect(&self, request: http::Request<()>) -> Result<WebSocketStream<ConnectStream>> {
+        let (mut parts, _) = request.into_parts();
+        if let Some(auth_header) = self.config.get_auth_header().await? {
+            parts.headers.insert(http::header::AUTHORIZATION, auth_header);
+        }
+        // Use the binary subprotocol v4.
+        // v4 is the current and sends JSON `metav1.Status` to `error` channel.
+        parts.headers.insert(
+            "sec-websocket-protocol",
+            "v4.channel.k8s.io".parse().expect("valid header value"),
```

Comment:
See [wsstream/conn.go](https://github.com/kubernetes/kubernetes/blob/0a839c6c3b9489716e216b4fc9dd2f9c17948c16/staging/src/k8s.io/apiserver/pkg/util/wsstream/conn.go#L34-L48) and [remotecommand/httpstream.go](https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/pkg/kubelet/cri/streaming/remotecommand/httpstream.go#L202-L204).

v4 sends JSON Status to the error channel (3).

Reply:
If there are no other official documents showing how the protocol works, I don't mind having these linked from the source code. Would be useful if/when we need to debug it further.